config/utils.ts

```diff
@@ -0,0 +1,447 @@
+
+import {
+  HUBSPOT_CONFIGURATION_FOLDER,
+  HUBSPOT_CONFIGURATION_FILE,
+  HUBSPOT_ACCOUNT_TYPES,
+  DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME,
+  ENVIRONMENT_VARIABLES,
+} from '../constants/config';
+import {
+  PERSONAL_ACCESS_KEY_AUTH_METHOD,
+  API_KEY_AUTH_METHOD,
+  OAUTH_AUTH_METHOD,
+  OAUTH_SCOPES,
+} from '../constants/auth';
+import { HubSpotConfig, DeprecatedHubSpotConfigFields } from '../types/Config';
+import { FileSystemError } from '../models/FileSystemError';
+import { logger } from '../lib/logger';
+import {
+  HubSpotConfigAccount,
+  OAuthConfigAccount,
+  AccountType,
+  TokenInfo,
+} from '../types/Accounts';
+import { getValidEnv } from '../lib/environment';
+import { getCwd } from '../lib/path';
+import { CMS_PUBLISH_MODE } from '../constants/files';
+import { i18n } from '../utils/lang';
+
+export function getGlobalConfigFilePath(): string {
+  return path.join(
+    os.homedir(),
+    HUBSPOT_CONFIGURATION_FOLDER,
+    HUBSPOT_CONFIGURATION_FILE
+  );
+}
+
+export function getLocalConfigFilePath(): string | null {
+  return findup([
+    DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME,
+    DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME.replace('.yml', '.yaml'),
+  ]);
+}
+
+export function getLocalConfigFileDefaultPath(): string {
+  return `${getCwd()}/${DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME}`;
+}
+
+export function getConfigPathEnvironmentVariables(): {
+  useEnvironmentConfig: boolean;
+  configFilePathFromEnvironment: string | undefined;
+} {
+  const configFilePathFromEnvironment =
+    process.env[ENVIRONMENT_VARIABLES.HUBSPOT_CONFIG_PATH];
+  const useEnvironmentConfig =
+    process.env[ENVIRONMENT_VARIABLES.USE_ENVIRONMENT_CONFIG] === 'true';
+
+  if (configFilePathFromEnvironment && useEnvironmentConfig) {
+    throw new Error(
+      i18n(
+        'config.utils.getConfigPathEnvironmentVariables.invalidEnvironmentVariables'
+      )
+    );
+  }
+
+  return {
+    configFilePathFromEnvironment,
+    useEnvironmentConfig,
+  };
+}
+
+export function readConfigFile(configPath: string): string {
+  let source = '';
+
+  try {
+    source = fs.readFileSync(configPath).toString();
+  } catch (err) {
+    throw new FileSystemError(
+      { cause: err },
+      {
+        filepath: configPath,
+        operation: 'read',
+      }
+    );
+  }
+
+  return source;
+}
+
+export function removeUndefinedFieldsFromConfigAccount<
+  T extends
+    | HubSpotConfigAccount
+    | Partial<HubSpotConfigAccount> = HubSpotConfigAccount,
+>(account: T): T {
+  Object.keys(account).forEach(k => {
+    const key = k as keyof T;
+    if (account[key] === undefined) {
+      delete account[key];
+    }
+  });
+
+  if ('auth' in account && account.auth) {
+    if (account.authType === OAUTH_AUTH_METHOD.value) {
+      Object.keys(account.auth).forEach(k => {
+        const key = k as keyof OAuthConfigAccount['auth'];
+        if (account.auth?.[key] === undefined) {
+          delete account.auth?.[key];
+        }
+      });
+    }
+
+    if (
```

Comment:
Going through these fields manually makes this function a bit more verbose than it needs to be, but keeps things type safe